### ISSUE TITLE
[rfc] ExternizerAttribute: Teach generator how to do P/Invokes from interface definitions.

### ISSUE THREAD

**Issue by dalexsoto** [xamarin/xamarin-macios#3452]
### Steps to Reproduce

Currently there is no way for the generator to produce [P/Invoke](http://www.mono-project.com/docs/advanced/pinvoke/) signatures.

### Expected Behavior

Generator must be able to produce P/Invoke definitions to avoid manual code. I must note that this is not intended to 100% remove manual code of P/Invokes but to be an additional tool to avoid it whenever possible.

### Potential Solution

Create a new attribute called `ExternizerAttribute` with the following definition:

```csharp
[AttributeUsage (AttributeTargets.Method, AllowMultiple = false)]
public class ExternizerAttribute : Attribute {

	public string LibName { get; set; }
	public string EntryPoint { get; set; }
	public bool ExternMethodOnly { get; set; }
	public Visibility ExternVisibility { get; set; } = Visibility.Disabled;
	public string CustomExternMethodName { get; set; }

	public ExternizerAttribute (string libName) => LibName = libName;
	public ExternizerAttribute (string libName, string entryPoint) : this (libName) => EntryPoint = entryPoint;
}
```

Property definitions:
* **LibName**: Native library to be opened, this will be mapped to `DllImport`'s `dllName` constructor parameter.
* **EntryPoint**: Native's library entry point to be called, this will be mapped to `DllImport`'s `EntryPoint` field.
* **ExternMethodOnly**: Instructs the generator to only create the extern signature and no wrapper method around it.
* **ExternVisibility**: Provides a way to customize extern's signature visibility, the default value is `Visibility.Disabled` which means that default visibility `private` will be used unless `ExternMethodOnly` is set to `true` this will turn visibility into `public`.
* **CustomExternMethodName**: Provides a way to customize extern method's name. By default extern method names will be suffixed with `Extern` unless `ExternMethodOnly` is set to `true` and `EntryPoint` is omitted.

A simple scenario of `ExternizerAttribute` would be as follows:

```csharp
[Externizer (Constants.MetalPerformanceShadersLibrary, "DoSomethingNative")]
NSSet DoSomething (MPSImage[] objs, [NullAllowed] MPSImage img, nint ratio);
```

And the output would be a wrapper method of the extern member that does casting and null checks:

```csharp
[DllImport ("/System/Library/Frameworks/MetalPerformanceShaders.framework/MetalPerformanceShaders", EntryPoint = "DoSomethingNative")]
static extern IntPtr DoSomethingExtern (IntPtr objs, IntPtr img, nint ratio);

public NSSet DoSomething (MPSImage[] objs, MPSImage img, nint ratio)
{
	if (objs == null)
		throw new ArgumentNullException ("objs");
	var nsa_objs = NSArray.FromNSObjects (objs);
	
	NSSet ret;
	ret =  Runtime.GetNSObject<NSSet> (DoSomethingExtern (nsa_objs.Handle, img == null ? IntPtr.Zero : img.Handle, ratio));
	nsa_objs.Dispose ();
	
	return ret;
}
```

Unfortunately `DllImport` cannot be reused because it cannot be applied to methods inside an interface definition like we do in our API definitions because the `extern` keyword is expected and some additional compiler checks. Also having our own attribute gives us the flexibility to add additional information needed to better drive the generator.

The following attributes would also be needed to be taken into account:

* [MarshalAs (UnmanagedType)] in return and parameters.
* [NullAllowed] for parameter checks.
* [return: Release] to allow to call `release` in the returned object if needed.
* [Advice]
* [iOS, Mac] and friends to write availability information.

### Notes

Again the scope of this attribute is not to get rid of manual code 100% but to help where possible for example it would enable the following scenarios:

* The default behaviour is meant for simple scenarios where null checks are needed before calling the unmanaged function.
* Generate the P/Invoke signature and use `[Wrap]` to do your thing.
* Expose the P/Invoke signature directly to the public API without manual code.
* C Arrays or variadic functions support is not in the scope of this attribute.

A small example where this attribute would be useful:

```c
typedef NSArray<MPSImage*>  MPSImageBatch;
NSUInteger MPSImageBatchIncrementReadCount( MPSImageBatch * __nonnull batch, NSInteger amount );
```

Would only require:
```csharp
[Externizer (Constants.MetalPerformanceShadersLibrary, "MPSImageBatchIncrementReadCount")]
nuint IncrementReadCount (MPSImage [] batch, nint amount);
```

💣💥🚀

**Comment by spouliot** [comment]
I like the idea but while it's fine **not** to support every possible cases, the proposal must support the most **common** cases (not just the ones we're trying to solve at this moment). Looking where our `DllImport` comes from it's clear that:

* it's uncommon to have p/invokes inside `NSObject`, most of them are inside `INativeObject` - that requires a bit of extra code (e.g. .ctors, Dispose pattern) but it would make things much easier (both for us and people doing bindings);

* unless they are exposed as static the first parameter of p/invokes are usually the handle of the type, which is not exposed in the public (generated) API. IOW your example above is uncommon for a non-static method;

* they must be able to use `.ctor (IntPtr, bool)` correctly when returning an `INativeObject` instance;

Look at `src/CoreGraphics/CGColorSpace.cs` and see how your current proposal would simplify things.

Minor points:

* `ExternVisibility` being `Disabled` meaning `Private` (which also exists in the enum) is confusing;

* `ExternVisibility` changing on `ExternMethodOnly` to public is confusing too;

* Not sure `CustomExternMethodName` is needed. The way `[DllImport]` works cover having different managed and native names;

* The name is funny but it's a verb, not an adjective, and that attribute does not externalize anything  - quite the reverse, it define an external to be used inside a library, so it's more an internalizer ;-) Maybe `LibImportAttribute` ? like you renamed `dllName` to `libName`

**Comment by dalexsoto** [comment]
> it's uncommon to have p/invokes inside NSObject, most of them are inside INativeObject - that requires a bit of extra code (e.g. .ctors, Dispose pattern) but it would make things much easier (both for us and people doing bindings);

> unless they are exposed as static the first parameter of p/invokes are usually the handle of the type, which is not exposed in the public (generated) API. IOW your example above is uncommon for a non-static method;

I love these ideas! ❤️

> they must be able to use .ctor (IntPtr, bool) correctly when returning an INativeObject instance;

This was the main idea behind supporting `[return: Release]` because we set `owns: false` when doing `GetNSObject` and `GetINativeObject` but yeah makes sense to bring this into `INativeObject` context.

> Look at src/CoreGraphics/CGColorSpace.cs and see how your current proposal would simplify things.

Will do!

---

> `ExternVisibility` being `Disabled` meaning Private (which also exists in the enum) is confusing;

This is because `Visibility` enum has `Disabled` value and can't change that nor I wanted to have yet another `Visibility` enum.  

```csharp
public Visibility ExternVisibility { get; set; } = Visibility.Disabled;
```

`ExternVisibility` is not part of the constructors and it is meant to be used if you really need to customize the extern signature visibility so making `ExternVisibility` having `Disabled` by default means that you will stick to the default behavior.

> ExternVisibility changing on ExternMethodOnly to public is confusing too;

`ExternVisibility` won't change, what will change is the visibility of the generated extern method, this is because if you set `ExternMethodOnly` to `true` means that no `public` wrapper method will be generated so the extern method might need to be `public` if you want to fully control visibility just use `ExternVisibility`.

> Not sure CustomExternMethodName is needed. The way [DllImport] works cover having different managed and native names

`CustomExternMethodName` is there to customize the managed name of the extern generated method, maybe do not want the `Extern` suffix? This allows more flexibility on the managed name.

> The name is funny but it's a verb, not an adjective, and that attribute does not externalize anything - quite the reverse, it define an external to be used inside a library, so it's more an internalizer ;-) Maybe LibImportAttribute ? like you renamed dllName to libName

`LibImportAttribute` sounds good to me 🙂 open to name suggestions

**Comment by rolfbjarne** [comment]
What about `[BindPInvokeAttribute]`?

Otherwise I think it's good :smile:

👍

**Comment by dalexsoto** [comment]
Added property to proposed attribute:

IsFirstParamHandle: Instructs the generator to create the P/Invoke signature with an extra parameter placed at the beginning that will be used to pass the Handle property of the container instance object.

This is basically to play nice with the generated class of [NativeObject](https://github.com/xamarin/xamarin-macios/issues/3582) and be able to generate quicker bindings.

**Comment by rolfbjarne** [comment]
@dalexsoto can't you use the `[Static]` attribute (or lack thereof) to do the same?

```csharp
[NativeObject]
interface NFoo {
    [Externizer (Constants.MetalPerformanceShadersLibrary, "DoSomethingNative")]
    NSSet DoSomething (IntPtr obj, MPSImage[] objs, [NullAllowed] MPSImage img, nint ratio);
}
```

This would generate an instance method on the `NFoo` class, where the first argument to the P/Invoke is the handle of the native object.

OTOH this:

```csharp
[NativeObject]
interface NFoo {
    [Static]
    [Externizer (Constants.MetalPerformanceShadersLibrary, "DoSomethingNative")]
    NSSet DoSomething (IntPtr obj, MPSImage[] objs, [NullAllowed] MPSImage img, nint ratio);
}
```

would generate a static P/Invoke that exposes the first IntPtr argument as well.

**Comment by dalexsoto** [comment]
Yeah @rolfbjarne maybe it can be a combo of [Static] && [NativeObject]

|             | [NativeObject]          | No [NativeObject]           |
|-------------|-------------------------|-----------------------------|
| [Static]    | first arg **is** handle | first arg is **not** handle |
| No [Static] | first arg **is** handle | first arg is **not** handle |

OTOH I think `IsFirstParamHandle` is more explicit on the intent, and also the other thing we might have not considered are the create methods these ones do not take the handle as the first param 🤔

**Comment by rolfbjarne** [comment]
I don't understand how the api definition would look with your suggestion, can you add an example?

The Create methods would be `[Static]`, and as such would not take the handle as the first param.

**Comment by migueldeicaza** [comment]
I like the general idea, but as Sebastien points out, Externizer is the wrong name.

I would rather go with `[PInvoke]`.   

I feel that we could make the library implicit in a given type, so that all P/Invokes in a given class by default use that, without having to specify it for every instance.

ExternVisibility is sausage making, if you do not want it to be public, use the existing attribute [Internal].   I rather not have another way of saying the same thing.

Without any glue code generated, I fail to see why I should use this capability, so [ExternMethodOnly], in that case, you should just go and write the P/Invoke by hand, it is likely shorter than this.

**Comment by dalexsoto** [comment]
Yeah I think this needs a little more of 🤔 💭 when working with `[NativeObject]` since I forgot that most of the time you get a status as return of the Create methods instead of the handle of the actual object like in VideoToolbox where the last parameter is an `out`

```csharp
[DllImport (Constants.VideoToolboxLibrary)]
extern static /* OSStatus */ VTStatus VTMultiPassStorageCreate (
	/* CFAllocatorRef */IntPtr allocator, /* can be null */
	/* CFURLRef */ IntPtr fileUrl, /* can be null */
	/* CMTimeRange */ CMTimeRange timeRange, /* can be kCMTimeRangeInvalid */
	/* CFDictionaryRef */ IntPtr options, /* can be null */
	/* VTMultiPassStorageRef */ out IntPtr multiPassStorageOut);
```

For the scenario where p/Invokes are just helper methods the current proposal works fine for example the [`VNUtils`](https://github.com/xamarin/xamarin-macios/blob/master/src/Vision/VNUtils.cs) and [`MTProfessionalVideoWorkflow`](https://github.com/xamarin/xamarin-macios/blob/master/src/MediaToolbox/MTProfessionalVideoWorkflow.cs) classes could be generated this way, actually @migueldeicaza answering your comment

> Without any glue code generated, I fail to see why I should use this capability, so [ExternMethodOnly], in that case, you should just go and write the P/Invoke by hand, it is likely shorter than this.

I think [`MTProfessionalVideoWorkflow`](https://github.com/xamarin/xamarin-macios/blob/master/src/MediaToolbox/MTProfessionalVideoWorkflow.cs) is a good scenario where you just would want the public p/invoke generated, this would allow us to kill this extra cs file and also would remove the need to track it inside our [frameworks.sources](https://github.com/xamarin/xamarin-macios/blob/0dd292f3d1e4d67cb2c60eba1dd84e5a613475b4/src/frameworks.sources#L972) file.

@rolfbjarne about your comment yeah, my mistake here is the updated matrix:

|             | [NativeObject]          | No [NativeObject]           |
|-------------|-------------------------|-----------------------------|
| [Static]    | first arg **is** handle | first arg is **not** handle |
| No [Static] | first arg is **not** handle~~ | first arg is **not** handle |

about one example of the Api definition without `[NativeObject]` would be:

 ```csharp
[Static]
interface MTProfessionalVideoWorkflow {

	[Mac (10,10)]
	[Static]
	[Externizer (Constants.MediaToolboxLibrary, "MTRegisterProfessionalVideoWorkflowFormatReaders", ExternMethodOnly = true)]
	void RegisterFormatReaders ();
}
```

This should generate something like this

```csharp
public static partial class MTProfessionalVideoWorkflow {
	[Introduced (PlatformName.MacOSX, 10,10, PlatformArchitecture.All)]
	[DllImport ("/System/Library/Frameworks/MediaToolbox.framework/MediaToolbox", EntryPoint = "MTRegisterProfessionalVideoWorkflowFormatReaders")]
	public static extern void RegisterFormatReaders ();
}
```

The same would apply to actual NSObjects where there is a `[BaseType]` over the interface, the extern and the glue method would just be generated and inlined there.

About `[NativeObject]` yes I need to think a little more about it, maybe we can split this proposal into two? With and without `[NativeObject]` support.

**Comment by rolfbjarne** [comment]
@dalexsoto I think the `[Static]` and no `[Static]` should be reversed: when no `[Static]` (and inside a `[NativeObject]`), it's like an instance method where the first argument is the instance itself.

**Comment by dalexsoto** [comment]
@rolfbjarne yes, had those backwards in my matrix, I fixed my above comment :D thanks 👍
